### PR TITLE
Patient app: Patient record download API

### PR DESCRIPTION
**Pull request by timcheadle**
**Story card:** https://www.pivotaltracker.com/story/show/172298466

## Because

The patient app needs to download the entire patient record for a single patient.

## This addresses

- [x] Authorizing API access using an access token
- [x] Rendering the patient record as a compiled JSON response

Response should include:
- [x] Patient ID
- [x] Patient demographics
- [x] Patient address
- [x] Patient phone numbers
- [x] Scanned BP passport (including short code)
- [x] Registration facility details
- [x] Medical history answers
- [x] Blood pressures
- [x] Blood sugars - 4c55357
- [x] Latest medications - 88fd683, 3ecfbaa
- [x] Appointments

## Sample response payload:

```json
{
  "patient": {
    "id": "4affcd53-8fbb-4341-b08d-c61580318350",
    "full_name": "Kiran Shreya",
    "age": 35,
    "gender": "transgender",
    "status": "active",
    "recorded_at": "2019-05-13T20:10:43.352Z",
    "reminder_consent": "granted",
    "phone_numbers": [
      {
        "id": "621c8573-c0b4-4d23-ad8e-cbf3695187e7",
        "number": "8462258734"
      }
    ],
    "address": {
      "id": "097c35bc-0e83-4053-9814-be722abfb208",
      "street_address": "215, 4th Main 3rd Ph, JP Nagar",
      "village_or_colony": "Eagle Square",
      "district": "Thiruvananthapuram",
      "zone": "Royal Square",
      "state": "Himachal Pradesh",
      "country": "Anguilla",
      "pin": "541543"
    },
    "registration_facility": {
      "id": "9dc3e56f-0866-4268-a3f3-a12ac3755480",
      "name": "DH Malkharoda",
      "street_address": "Gr Floor, Plot No 260, Mehar Bldg, Sector 28, Vashi",
      "village_or_colony": "Malkharoda",
      "district": "Majuli",
      "state": "Maharashtra",
      "country": "India",
      "pin": "936166"
    },
    "blood_pressures": [
      {
        "systolic": 191,
        "diastolic": 63,
        "recorded_at": "2019-11-05T07:46:50.654Z",
        "facility": {
          "id": "9dc3e56f-0866-4268-a3f3-a12ac3755480",
          "name": "DH Malkharoda",
          "street_address": "Gr Floor, Plot No 260, Mehar Bldg, Sector 28, Vashi",
          "village_or_colony": "Malkharoda",
          "district": "Majuli",
          "state": "Maharashtra",
          "country": "India",
          "pin": "936166"
        }
      },
      {
        "systolic": 146,
        "diastolic": 84,
        "recorded_at": "2019-07-29T15:42:55.399Z",
        "facility": {
          "id": "9dc3e56f-0866-4268-a3f3-a12ac3755480",
          "name": "DH Malkharoda",
          "street_address": "Gr Floor, Plot No 260, Mehar Bldg, Sector 28, Vashi",
          "village_or_colony": "Malkharoda",
          "district": "Majuli",
          "state": "Maharashtra",
          "country": "India",
          "pin": "936166"
        }
      }
    ],
    "blood_sugars": [
      {
        "blood_sugar_value": "85.0",
        "blood_sugar_type": "random",
        "recorded_at": "2020-04-20T18:04:55.378Z",
        "facility": {
          "id": "9dc3e56f-0866-4268-a3f3-a12ac3755480",
          "name": "DH Malkharoda",
          "street_address": "Gr Floor, Plot No 260, Mehar Bldg, Sector 28, Vashi",
          "village_or_colony": "Malkharoda",
          "district": "Majuli",
          "state": "Maharashtra",
          "country": "India",
          "pin": "936166"
        }
      },
      {
        "blood_sugar_value": "95.0",
        "blood_sugar_type": "fasting",
        "recorded_at": "2020-04-17T18:06:32.535Z",
        "facility": {
          "id": "9dc3e56f-0866-4268-a3f3-a12ac3755480",
          "name": "DH Malkharoda",
          "street_address": "Gr Floor, Plot No 260, Mehar Bldg, Sector 28, Vashi",
          "village_or_colony": "Malkharoda",
          "district": "Majuli",
          "state": "Maharashtra",
          "country": "India",
          "pin": "936166"
        }
      }
    ],
    "appointments": [
      {
        "scheduled_date": "2020-02-13",
        "status": "scheduled",
        "facility": {
          "id": "9dc3e56f-0866-4268-a3f3-a12ac3755480",
          "name": "DH Malkharoda",
          "street_address": "Gr Floor, Plot No 260, Mehar Bldg, Sector 28, Vashi",
          "village_or_colony": "Malkharoda",
          "district": "Majuli",
          "state": "Maharashtra",
          "country": "India",
          "pin": "936166"
        }
      },
      {
        "scheduled_date": "2020-03-13",
        "status": "scheduled",
        "facility": {
          "id": "9dc3e56f-0866-4268-a3f3-a12ac3755480",
          "name": "DH Malkharoda",
          "street_address": "Gr Floor, Plot No 260, Mehar Bldg, Sector 28, Vashi",
          "village_or_colony": "Malkharoda",
          "district": "Majuli",
          "state": "Maharashtra",
          "country": "India",
          "pin": "936166"
        }
      }
    ],
    "medications": [
      {
        "name": "Amlodipine",
        "dosage": "5mg",
        "rxnorm_code": null,
        "is_protocol_drug": true
      },
      {
        "name": "Telmesartin",
        "dosage": "40mg",
        "rxnorm_code": null,
        "is_protocol_drug": true
      }
    ]
  }
}
```

### CODE REVIEW

**Comment by timcheadle** [comment]
@harimohanraj89 we need to add details about the currently scanned BP passport to the patient payload. The easiest way to do this would be to pull the passport from the matching `PassportAuthentication` record and shove the UUID and shortcode into the payload. Can we pair on implementing this?

**Comment by harimohanraj89** [comment]
Sounds like a plan @timcheadle . Another major implementation note on this PR: I was looking at how we handle multiple BP passports, and look up access tokens, etc. There was a sort of unpredictable element where a patient could get a second access token with a second BP passport on a second device, but the device wouldn't be able to authenticate itself because the access-token-finding would only find the first token, not the second one.

Spiking some options also revealed that the Passport Auth -> Patient relationship may be redundant. We can instead associate patients through the BP Passport record.

I've removed the extra association in b3f7aff, added a `patient.access_tokens` convenience method in ec22b9c, and authenticated patients against _all_ of their valid access tokens in 2d4af12

**Comment by harimohanraj89** [comment]
@timcheadle As a first take, I've added the passport deets to the `login` response, as it's tied with the concept of "which passport did I just authenticate with?". - af49502

This way, the app can also immediately stash the shortcode against the passport after sign in, and fetch patient info lazily.

File: app/controllers/api/v4/patient_controller.rb [hunk 1]
```diff
@@ -1,4 +1,7 @@
 class Api::V4::PatientController < PatientAPIController
```
Comment (vkrmis) [inline app/controllers/api/v4/patient_controller.rb:1]: I see the controller name is the singular form, should we disambiguate it more from a potential V4 `PatientsController`? Maybe `PatientAppController`?

Comment (timcheadle) [reply app/controllers/api/v4/patient_controller.rb:1]: No; this is standard Rails convention for a singular vs. plural route

File: app/controllers/api/v4/patient_controller.rb [hunk 2]
```diff
@@ -7,10 +10,7 @@ def activate
     patient  = passport&.patient
     return head :not_found unless patient.present? && patient.latest_mobile_number.present?
```
Comment (vkrmis) [inline app/controllers/api/v4/patient_controller.rb:11]: Should `patient.latest_mobile_number.present?` throw a different error than `not_found`? That way the patient can be informed that they need to ask their nurse/doctor to update their number.

Comment (timcheadle) [reply app/controllers/api/v4/patient_controller.rb:11]: We discussed this and we may do that later. For now we're keeping it simple.

**Pull request merged**